// eyeglass/lib/util/perf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var LRU = require("lru-cache");
function resetGlobalCaches() {
    realpathSync.resetCache();
    existsSync.resetCache();
}
exports.resetGlobalCaches = resetGlobalCaches;
/* cache real path locations so we don't hit the filesystem
 more than once for the same path every 5 min. We cap this
 cache at ~1MB of memory.*/
var realpathCache = new LRU({
    max: 1024 * 1024,
    length: function (value, key) { return value.length + key.length; },
    maxAge: 5 * 60 * 1000 /* 5 min */,
});
function realpathSync(path) {
    var value = realpathCache.get(path);
    if (value) {
        return value;
    }
    value = fs.realpathSync(path);
    realpathCache.set(path, value);
    return value;
}
exports.realpathSync = realpathSync;
realpathSync.resetCache = function () {
    realpathCache.reset();
};
var existsCache = Object.create(null);
function existsSync(path) {
    var result = existsCache[path];
    if (result === true || result === false) {
        return result;
    }
    else {
        result = fs.existsSync(path);
        existsCache[path] = result;
        return result;
    }
}
exports.existsSync = existsSync;
existsSync.resetCache = function () {
    existsCache = Object.create(null);
};
